pandas/core/apply.py

```diff
@@ -280,6 +280,7 @@ def apply_standard(self):
             and not self.dtypes.apply(is_extension_array_dtype).any()
             # Disallow complex_internals since libreduction shortcut
             #  cannot handle MultiIndex
+            and not self.dtypes.apply(lambda x: x.kind in ["m", "M"]).any()
```

Comment:
can you also update the comment here on why this special case is needed?

Reply:
should this be: ```self.dtypes.apply(is_extension_array_dtype).any()```?

Reply:
thats up on L280

Reply:
Has this a performance implication for datetime64[ns] ? Because before this was taking the libreduction path, and now not anymore?

(btw, the comment "Disallow dtypes that have blocks backed by EAs" is not fully correct, as DatetimeBlock is still backed by ndarray no?)

Reply:
Looking for either a more accurate comment or a way around this.  Getting rid of this new check I get one test failure locally:

```
self = <pandas.tests.frame.test_analytics.TestDataFrameAnalytics object at 0x127a63050>
float_frame_with_na =                    A         B         C         D
ew6ipdVXcu -1.978064 -0.908831 -0.752288  0.223983
elVr6w6e8F  0.47....447613 -2.139044
322RdJJZNU  1.993131 -0.053709 -2.386576  0.199983
nzUStPA2lc -0.402711 -0.593735  0.679934 -1.681615
mixed_float_frame =                    A         B         C         D
TdfPTU8h12  0.336661  0.578183 -0.099304 -1.618718
dHJdHvmOic -0.01....558594  1.554425
XvSW0hK7vE -0.552191  0.844111 -0.937012 -0.398246
UR9qvSyscK -0.856967  1.425748 -1.767578 -0.524487

    def test_stat_op_calc(self, float_frame_with_na, mixed_float_frame):

[...]
pandas/tests/frame/test_analytics.py:72: in assert_stat_op_calc
    result = getattr(df, opname)()
pandas/core/frame.py:8075: in nunique
    return self.apply(Series.nunique, axis=axis, dropna=dropna)
pandas/core/frame.py:6946: in apply
    return op.get_result()
pandas/core/apply.py:187: in get_result
    return self.apply_standard()
pandas/core/apply.py:300: in apply_standard
    values, self.f, axis=self.axis, dummy=dummy, labels=labels
pandas/_libs/reduction.pyx:617: in pandas._libs.reduction.compute_reduction
    return reducer.get_result()
pandas/_libs/reduction.pyx:127: in pandas._libs.reduction.Reducer.get_result
    res = self.f(cached_typ)
pandas/core/apply.py:114: in f
    return func(x, *args, **kwds)
pandas/core/base.py:1311: in nunique
    uniqs = self.unique()
pandas/core/series.py:1784: in unique
    result = super().unique()
pandas/core/base.py:1267: in unique
    result = values.unique()
pandas/core/arrays/datetimelike.py:650: in unique
    result = unique1d(self.asi8)
pandas/core/arrays/datetimelike.py:454: in asi8
    return self._data.view("i8")
[...]
        if newtype.hasobject or oldtype.hasobject:
>           raise TypeError("Cannot change data-type for object array.")
E           TypeError: Cannot change data-type for object array.
```

IIUC inside libreducer we are setting _index_data, which indirectly alters the _data on a TimedeltaArray, so when we try to call `self._data.view("i8")` it chokes.  That more or less make sense.

what doesnt make sense is why a call on float_frame_with_na is doing _anything_ with a timedelta64 dtype

Reply:
That's indeed strange ...

Did you check if there is a performance difference?

Reply:
>>> Did you check if there is a performance difference?

Non-zero but not huge:

```
dti = pd.date_range("2016-01-01", periods=10**6, freq="S")
tdi = dti - dti
ser = pd.Series(tdi)
df = ser.to_frame()

%timeit df.apply(lambda x: x.median())
23.7 ms ± 251 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <-- master
24.2 ms ± 426 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <-- PR
```

consistent across several timeit runs about 2% slowdown